src/ai/backend/storage/xfs/__init__.py

```diff
@@ -1,96 +1,91 @@
-    if err:
-        raise ExecutionError(err.decode())
-    return out.decode()
-
-
-class XfsVolume(BaseVolume):
-    loop: asyncio.AbstractEventLoop
-    registry: Dict[str, int]
-    project_id_pool: List[int]
-
-    async def init(self, uid=None, gid=None, loop=None) -> None:
-        self.registry = {}
-        self.project_id_pool = []
-        if uid is not None:
-            self.uid = uid
-        else:
-            self.uid = os.getuid()
-        if gid is not None:
-            self.gid = gid
-        else:
-            self.gid = os.getgid()
-        self.loop = loop or asyncio.get_running_loop()
-
-        if os.path.isfile("/etc/projid"):
-            raw_projid = await read_file(self.loop, "/etc/projid")
+        # TODO: how to handle if /etc/proj* files are deleted by external reason?
+        # TODO: do we need to use /etc/proj* files to enlist the project information?
+        if self.file_projid.is_file():
+            project_id_pool = []
+            self.name_id_map = {}
+            loop = asyncio.get_running_loop()
+            raw_projid = await loop.run_in_executor(None, _read_projid_file)
             for line in raw_projid.splitlines():
                 proj_name, proj_id = line.split(":")[:2]
-                self.project_id_pool.append(int(proj_id))
-                self.registry[proj_name] = int(proj_id)
-            self.project_id_pool = sorted(self.project_id_pool)
+                project_id_pool.append(int(proj_id))
+                self.name_id_map[UUID(proj_name)] = int(proj_id)
+            self.project_id_pool = sorted(project_id_pool)
         else:
-            await self.loop.run_in_executor(None, lambda: Path("/etc/projid").touch())
-
-        if not os.path.isfile("/etc/projects"):
-            await self.loop.run_in_executor(None, lambda: Path("/etc/projects").touch())
+            await run(["sudo", "touch", self.file_projid])
+        if not Path(self.file_projects).is_file():
+            await run(["sudo", "touch", self.file_projects])
 
-    # ----- volume opeartions -----
-    async def create_vfolder(
+    async def add_project_entry(
         self,
+        *,
         vfid: UUID,
-        options: VFolderCreationOptions = None,
+        quota: int,
+        project_id: int = None,
     ) -> None:
-        if str(vfid) in self.registry.keys():
-            raise VFolderCreationError("VFolder id {} already exists".format(str(vfid)))
+        vfpath = self.backend.mangle_vfpath(vfid)
+        if project_id is None:
+            project_id = self.get_project_id()
+        await run(
+            [
+                "sudo",
+                "sh",
+                "-c",
+                f"echo '{project_id}:{vfpath}' >> {self.file_projects}",
+            ],
+        )
+        await run(
+            [
+                "sudo",
+                "sh",
+                "-c",
+                f"echo '{str(vfid)}:{project_id}' >> {self.file_projid}",
+            ],
+        )
```

Comment:
It is better to just use a temporary file and move it using `sudo mv`, to avoid potential shell injection here, following the same rationale to change `run()` to use `create_subprocess_exec()` instead of `create_subprocess_shell()`. @adrysn If you have time, I'd suggest to fix it like so.

Ideally it is better to run the storage proxy service as root instead of relying on sudo-capability as `sudo` may get affected by service-specific environment setup (e.g., difference of `PATH` variable, etc., when launched via systemd). We need to carefully document the prerequisite for using this XFS backend, at least in README.